src/NHibernate/Engine/Query/QueryPlanCache.cs

```diff
@@ -95,26 +95,30 @@ public IQueryExpressionPlan GetHQLQueryPlan(IQueryExpression queryExpression, bo
 			return plan;
 		}
 
-		public FilterQueryPlan GetFilterQueryPlan(string filterString, string collectionRole, bool shallow, IDictionary<string, IFilter> enabledFilters)
+		public IQueryExpressionPlan GetFilterQueryPlan(string filterString, string collectionRole, bool shallow, IDictionary<string, IFilter> enabledFilters)
 		{
-			var key = new FilterQueryPlanKey(filterString, collectionRole, shallow, enabledFilters);
-			var plan = (FilterQueryPlan) planCache[key];
+			return GetFilterQueryPlan(new StringQueryExpression(filterString), collectionRole, shallow, enabledFilters);
+		}
+
+		public IQueryExpressionPlan GetFilterQueryPlan(IQueryExpression queryExpression, string collectionRole, bool shallow, IDictionary<string, IFilter> enabledFilters)
+		{
+			var key = new FilterQueryPlanKey(queryExpression.Key, collectionRole, shallow, enabledFilters);
+			var plan = (IQueryExpressionPlan) planCache[key];
 
 			if (plan == null)
 			{
 				if (log.IsDebugEnabled)
 				{
-					log.Debug("unable to locate collection-filter query plan in cache; generating (" + collectionRole + " : "
-							  + filterString + ")");
+					log.Debug(string.Format("unable to locate collection-filter query plan in cache; generating ({0} : {1})", collectionRole, queryExpression.Key));
```

Comment:
>I guess at the time of this change, $"" formatting was not available!
True. However, It seems that it shall be changed to `log.DebugFormat(..)`

>`if (log.IsDebugEnabled)`

It seems just to avoid doing string concatenations and win some milliseconds and object allocations (in some cases it's actually a huge win)  

Haven't found anything on C#, so the links are about Java, but internals are almost the same, so 

- http://stackoverflow.com/questions/963492/in-log4j-does-checking-isdebugenabled-before-logging-improve-performance
- https://www.igorkromin.net/index.php/2015/05/12/are-guard-statements-like-isdebugenabled-necessary-when-using-log4j/
- http://learnertobeginner.blogspot.co.nz/2010/06/how-to-use-isdebugenabled-effectly.html